src/probe/probe.c

```diff
@@ -15,42 +31,1072 @@
+		for (j = 0; j < CONFIG_PROBE_DMA_MAX; j++) {
+			if (_probe->inject_dma[j].stream_tag == stream_tag[i]) {
+				err = probe_dma_deinit(&_probe->inject_dma[j]);
+				if (err < 0)
+					return err;
+			}
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Copy data to probe buffer and update buffer pointers.
+ * \param[out] probe DMA buffer.
+ * \param[in] data pointer.
+ * \param[in] size.
+ * \return 0 on success, error code otherwise.
+ */
+static int copy_to_pbuffer(struct probe_dma_buf *pbuf, void *data, uint32_t bytes)
+{
+	uint32_t head;
+	uint32_t tail;
+
+	if (bytes == 0)
+		return 0;
+
+	/* check if it will not exceed end_addr */
+	if ((char *)pbuf->end_addr - (char *)pbuf->w_ptr < bytes) {
+		head = (char *)pbuf->end_addr - (char *)pbuf->w_ptr;
+		tail = bytes - head;
+	} else {
+		head = bytes;
+		tail = 0;
+	}
+
+	/* copy data to probe buffer */
+	if (memcpy_s((void *)pbuf->w_ptr, pbuf->end_addr - pbuf->w_ptr, data, head)) {
+		trace_probe_error("copy_to_pbuffer() error: memcpy_s() failed");
+		return -EINVAL;
+	}
+	dcache_writeback_region((void *)pbuf->w_ptr, head);
+
+	/* buffer ended so needs to do a second copy */
+	if (tail) {
+		pbuf->w_ptr = pbuf->addr;
+		if (memcpy_s((void *)pbuf->w_ptr, (char *)pbuf->end_addr - (char *)pbuf->w_ptr,
+			     (char *)data + head, tail)) {
+			trace_probe_error("copy_to_pbuffer() error: memcpy_s() failed");
+			return -EINVAL;
+		}
+		dcache_writeback_region((void *)pbuf->w_ptr, tail);
+		pbuf->w_ptr = pbuf->w_ptr + tail;
+	} else {
+		pbuf->w_ptr = pbuf->w_ptr + head;
+	}
+
+	pbuf->avail = (uintptr_t)pbuf->avail + bytes;
+
+	return 0;
+}
+
+/**
+ * \brief Copy data from probe buffer and update buffer pointers.
+ * \param[out] probe DMA buffer.
+ * \param[out] data pointer.
+ * \param[in] size.
+ * \return 0 on success, error code otherwise.
+ */
+static int copy_from_pbuffer(struct probe_dma_buf *pbuf, void *data, uint32_t bytes)
+{
+	uint32_t head;
+	uint32_t tail;
+
+	if (bytes == 0)
+		return 0;
+	/* not enough data available so set it to 0 */
+	if (pbuf->avail < bytes) {
+		memset(data, 0, bytes);
+		return 0;
+	}
+
+	/* check if memcpy needs to be divided into two stages */
+	if (pbuf->end_addr - pbuf->r_ptr < bytes) {
+		head = pbuf->end_addr - pbuf->r_ptr;
+		tail = bytes - head;
+	} else {
+		head = bytes;
+		tail = 0;
+	}
+
+	/* data from DMA so invalidate it */
+	dcache_invalidate_region((void *)pbuf->r_ptr, head);
+	if (memcpy_s(data, bytes, (void *)pbuf->r_ptr, head)) {
+		trace_probe_error("copy_from_pbuffer() error: memcpy_s() failed");
+		return -EINVAL;
+	}
+
+	/* second stage copy */
+	if (tail) {
+		/* starting from the beginning of the buffer */
+		pbuf->r_ptr = pbuf->addr;
+		dcache_invalidate_region((void *)pbuf->r_ptr, tail);
+		if (memcpy_s((char *)data + head, tail, (void *)pbuf->r_ptr, tail)) {
+			trace_probe_error("copy_from_pbuffer() error: memcpy_s() failed");
+			return -EINVAL;
+		}
+		pbuf->r_ptr = pbuf->r_ptr + tail;
+	} else {
+		pbuf->r_ptr = pbuf->r_ptr + head;
+	}
```

Comment:
I feel like there is a lot of repetition of raw buffer management code. Is there no way to refactor it, possibly sharing some code with audio buffers (streams) as well?

Reply:
I dont think so, we have 2 simple functions for data copying

Reply:
The functions are anything but simple. This type of buffer code is also available in the audio stream, and thus we have 3 instances in the whole project. Because it's easy to make subtle mistakes (which may slip review as well) I strongly believe there should be only one instance of this kind of code, usable by all the circular buffers in the project. Perhaps a circular_buffer structure and functions which will be used by everyone?

@tlauda @lgirdwood What's your take on this one?